### PR TITLE
fix(compiler): check entry component metadata

### PR DESCRIPTION
**Pull request by bowenni**
Throw a proper error when a non-component is added to module's bootstrap list.
This avoids adding non-components to entry components and avoids printing stack traces.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 1]
```diff
@@ -949,6 +950,10 @@ export class CompileMetadataResolver {
         return {componentType: dirType, componentFactory: dirSummary.componentFactory};
       }
     }
+
+    if (throwIfNotFound) {
+      throw new SyntaxError(`${dirType.name} is not an entry component.`);
```
Comment (tbosch) [inline modules/@angular/compiler/src/metadata_resolver.ts]: Or `... can not be used as an entry component`.

File: modules/@angular/compiler/test/metadata_resolver_spec.ts [hunk 2]
```diff
@@ -325,6 +325,60 @@ export function main() {
...
+
+         @NgModule({declarations: [], bootstrap: [SimpleService]})
+         class ModuleWithServiceInBootstrap {
+         }
+
+         expect(() => resolver.getNgModuleMetadata(ModuleWithServiceInBootstrap))
+             .toThrowError(SyntaxError, `SimpleService is not an entry component.`);
+       }));
+
+    it(`should not throw an error when a Directive is added to module's bootstrap list`,
```
Comment (tbosch) [inline modules/@angular/compiler/test/metadata_resolver_spec.ts]: Actually it should throw in this case as well!

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 3]
```diff
@@ -930,15 +930,16 @@ export class CompileMetadataResolver {
...
       }
     });
     return components;
   }
 
-  private _getEntryComponentMetadata(dirType: any): cpl.CompileEntryComponentMetadata {
+  private _getEntryComponentMetadata(dirType: any, throwIfNotFound = true):
+      cpl.CompileEntryComponentMetadata {
     const dirMeta = this.getNonNormalizedDirectiveMetadata(dirType);
     if (dirMeta) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/metadata_resolver.ts]: check `DirMeta.isComponent` / `dirSummary.isComponent` to be truthy...

**Review by tbosch — APPROVED** [review]

**Comment by bowenni** [comment]
@tbosch PTAL

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**